tfjs-backend-webgpu/src/lrn_webgpu.ts

```diff
@@ -0,0 +1,148 @@
+}
+`;
+
+export class LRNProgram implements WebGPUProgram {
+  outputShape: number[] = [];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x'];
+  uniforms = 'radius : i32, bias : f32, alpha : f32, beta : f32,';
+  workgroupSize: [number, number, number] = [64, 1, 1];
+  size = true;
+
+  constructor(xShape: number[]) {
+    this.outputShape = xShape;
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+    this.shaderKey = 'lrn';
+  }
+
+  getUserCode(): string {
+    const userCode = `
+    ${main('index')} {
+      if (index < uniforms.size) {
+        let coords = getOutputCoords();
+        let b = coords[0];
+        let r = coords[1];
+        let c = coords[2];
+        let d = coords[3];
+
+        let x = getX(b, r, c, d);
+        var sum = 0.0;
+        for (var i = -uniforms.radius; i <= uniforms.radius; i = i + 1) {
+          let idx = d + i;
+          if (idx >= 0 && idx < uniforms.xShape[3]) {
+            let z = getX(b, r, c, idx);
+            sum = sum + z * z;
+          }
+        }
+        ${powOperatorSnippet}
+
+        setOutputAtIndex(index, x * powValue);
+      }
+    }
+  `;
+    return userCode;
+  }
+}
+
+export class LRNSharedProgram implements WebGPUProgram {
+  outputShape: number[] = [];
+  shaderKey: string;
+  dispatchLayout: {x: number[], y: number[], z: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x'];
+  uniforms = 'radius : i32, bias : f32, alpha : f32, beta : f32,';
+  workgroupSize: [number, number, number] = [256, 1, 1];
+  maxAllowRadius = 16;
+  elementsPerWorkgroup: number;
+
+  constructor(xShape: number[], radius: number) {
+    util.assert(
+        radius <= this.maxAllowRadius,
+        () => `Radius must be less than or equal to ${
+            this.maxAllowRadius}, current radius is ${radius}`);
+
+    this.outputShape = xShape;
+    // The reason why not using this.workgroupSize[0] + 2 * maxAllowRadius here
+    // is to make sure that there is only one time gobal memory access for each
+    // thread.
+    this.elementsPerWorkgroup = this.workgroupSize[0] - 2 * this.maxAllowRadius;
+    this.dispatchLayout = {x: [3], y: [2], z: [0, 1]};
+    this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape, [
+      this.elementsPerWorkgroup, this.workgroupSize[1], this.workgroupSize[2]
+    ]);
+    this.shaderKey = 'lrn_shared';
+  }
+
+  getUserCode(): string {
+    const userCode = `
+    var <workgroup>lrnSub: array<f32, ${this.workgroupSize[0]}>;
+    const elementsPerWorkgroup = ${this.elementsPerWorkgroup};
+    const maxAllowRadius = ${this.maxAllowRadius};
+
+    ${main()} {
+      let b = i32(globalId.z) / uniforms.xShape[1];
+      let r = i32(globalId.z) - b * uniforms.xShape[1];
+      let c = i32(globalId.y);
+      let d = i32(globalId.x);
```

Comment:
Can we directly use below snippet?
```
        let coords = getOutputCoords();
        let b = coords[0];
        let r = coords[1];
        let c = coords[2];
        let d = coords[3];
```

Reply:
> let coords = getOutputCoords();
>         let b = coords[0];
>         let r = coords[1];
>         let c = coords[2];
>         let d = coords[3];

@Jiajia, I had addressed all your comments except above one. The reason is that `d` coordinate is not the same as the usual one, it must subtract `2 * maxAllowRadius`  for every work group.